### ISSUE TITLE
Return const reference for EAbstractData metadata

### ISSUE THREAD

**Issue by bentsherman** [SystemsGenetics/ACE#73]
I know that returning a copy is intentional, but it should at least return a const reference in case a class just wants to read the metadata. Now that the import/export analytics use multiple work blocks, I have to get the metadata for every function call, which means that the metadata is being unnecessarily copied every time. For the extract analytic it actually made a huge performance difference. I can get around it for now by initializing the metadata variable once as a class member, but it is less elegant and a typical user will definitely expect to be able to access metadata without having to copy it.

**Comment by 4ctrl-alt-del** [comment]
Just to make sure I am understanding you correctly change these two protected methods in EAbstractData:
```
   EMetadata systemMeta() const;
   EMetadata meta() const;
```
To:
```
   const EMetadata& systemMeta() const;
   const EMetadata& meta() const;
```
Yes?

**Comment by bentsherman** [comment]
That's it!

**Comment by 4ctrl-alt-del** [comment]
Alright changed in commit e1b7ecc4bbeb42f7a8feaad3095d0c3b722752f0.

Let me know if this is what you wanted. Tested it and seems to work for me.

**Comment by bentsherman** [comment]
One more issue -- you need to apply the same change to `EMetaObject`, specifically to `operator[]` and `at()`, to also return a const reference instead of copy. I think `EMetaArray` is already correct.

**Comment by 4ctrl-alt-del** [comment]
That I cannot do. EMetaObject is basically a wrapper of QMap<QString,EMetadata>. I cannot provide a reference to an r-value.

**Comment by bentsherman** [comment]
Okay I see, looking at the QMap documentation. In that case, I was able to get a const reference using `QMap::find()` like this:
```
const EMetaArray& ExpressionMatrix::geneNames() const
{
   return meta().toObject().find("genes")->toArray();
}
```

However I had to change EMetaObject to inherit from QMap as public so that this method was accessible. Could we change EMetaObject this way, or at least add a wrapper method for `find()`? It seems to be the only way to get a const reference from a QMap.
